File: csharp/core/lwf_input.cs
/*
 * Copyright (C) 2012 GREE, Inc.
 * 
 * This software is provided 'as-is', without any express or implied
 * warranty.  In no event will the authors be held liable for any damages
 * arising from the use of this software.
 * 
 * Permission is granted to anyone to use this software for any purpose,
 * including commercial applications, and to alter it and redistribute it
 * freely, subject to the following restrictions:
 * 
 * 1. The origin of this software must not be misrepresented; you must not
 *    claim that you wrote the original software. If you use this software
 *    in a product, an acknowledgment in the product documentation would be
 *    appreciated but is not required.
 * 2. Altered source versions must be plainly marked as such, and must not be
 *    misrepresented as being the original software.
 * 3. This notice may not be removed or altered from any source distribution.
 */

namespace LWF {

public partial class LWF
{
	public Button InputPoint(int px, int py)
	{
		m_intercepted = false;

		if (!interactive)
			return null;

		float x = px;
		float y = py;

		m_pointX = x;
		m_pointY = y;

		bool found = false;
		for (Button button = m_buttonHead;
				button != null; button = button.buttonLink) {
			if (button.CheckHit(x, y)) {
				if (m_allowButtonList != null) {
					bool v;
					if (!m_allowButtonList.TryGetValue(
							button.instanceId, out v)) {
						if (m_interceptByNotAllowOrDenyButtons) {
							m_intercepted = true;
							break;
						} else {
							continue;
						}
					}
				} else if (m_denyButtonList != null) {
					bool v;
					if (m_denyButtonList.TryGetValue(
							button.instanceId, out v)) {
						if (m_interceptByNotAllowOrDenyButtons) {
							m_intercepted = true;
							break;
						} else {
							continue;
						}
					}
				}

				found = true;
				if (m_focus != button) {
					if (m_focus != null)
						m_focus.RollOut();
					m_focus = button;
					m_focus.RollOver();
				}
				break;
			}
		}
		if (!found && m_focus != null) {
			m_focus.RollOut();
			m_focus = null;
		}

		return m_focus;
	}

	public void InputPress()
	{
		if (!interactive)
			return;

		m_pressing = true;

		if (m_focus != null) {
			m_pressed = m_focus;
			m_focus.Press();
		}
	}

	public void InputRelease()
	{
		if (!interactive)
			return;

		m_pressing = false;

		if (m_focus != null && m_pressed == m_focus) {
			m_focus.Release();
			m_pressed = null;
		}
	}

	public void InputKeyPress(int code)
	{
		if (!interactive)
			return;

		for (Button button = m_buttonHead;
				button != null; button = button.buttonLink) {
			button.KeyPress(code);
		}
	}
}

}	// namespace LWF

File: cplusplus/core/lwf_input.cpp
/*
 * Copyright (C) 2013 GREE, Inc.
 * 
 * This software is provided 'as-is', without any express or implied
 * warranty.  In no event will the authors be held liable for any damages
 * arising from the use of this software.
 * 
 * Permission is granted to anyone to use this software for any purpose,
 * including commercial applications, and to alter it and redistribute it
 * freely, subject to the following restrictions:
 * 
 * 1. The origin of this software must not be misrepresented; you must not
 *    claim that you wrote the original software. If you use this software
 *    in a product, an acknowledgment in the product documentation would be
 *    appreciated but is not required.
 * 2. Altered source versions must be plainly marked as such, and must not be
 *    misrepresented as being the original software.
 * 3. This notice may not be removed or altered from any source distribution.
 */

#include "lwf_core.h"
#include "lwf_button.h"

namespace LWF {

Button *LWF::InputPoint(int px, int py)
{
	intercepted = false;

	if (!interactive)
		return 0;

	float x = px;
	float y = py;

	pointX = x;
	pointY = y;

	bool found = false;
	for (Button *button = buttonHead; button; button = button->buttonLink) {
		if (button->CheckHit(x, y)) {
			if (!m_allowButtonList.empty()) {
				if (m_allowButtonList.find(button->instanceId) ==
						m_allowButtonList.end()) {
					if (interceptByNotAllowOrDenyButtons) {
						intercepted = true;
						break;
					} else {
						continue;
					}
				}
			} else if (!m_denyButtonList.empty()) {
				if (m_denyButtonList.find(button->instanceId) !=
						m_denyButtonList.end()) {
					if (interceptByNotAllowOrDenyButtons) {
						intercepted = true;
						break;
					} else {
						continue;
					}
				}
			}

			found = true;
			if (focus != button) {
				if (focus)
					focus->RollOut();
				focus = button;
				focus->RollOver();
			}
			break;
		}
	}
	if (!found && focus) {
		focus->RollOut();
		focus = 0;
	}

	return focus;
}

void LWF::InputPress()
{
	if (!interactive)
		return;

	pressing = true;

	if (focus) {
		pressed = focus;
		focus->Press();
	}
}

void LWF::InputRelease()
{
	if (!interactive)
		return;

	pressing = false;

	if (focus && pressed == focus) {
		focus->Release();
		pressed = 0;
	}
}

void LWF::InputKeyPress(int code)
{
	if (!interactive)
		return;

	for (Button *button = buttonHead; button; button = button->buttonLink) {
		button->KeyPress(code);
	}
}

}	// namespace LWF
